packages/b/bgfx/rules/shaders.lua

```diff
@@ -89,13 +77,39 @@ rule("shaders")
             }
         }
 
+        local shader_type
+        if fileconfig and fileconfig.type then
+            if table.contains(bgfx_types, fileconfig.type) then
+                shader_type = fileconfig.type
+            else
+                raise("unsupported shader type " .. fileconfig.type)
+            end
+        elseif shaderfile:match("^vs_.*%.sc$") or shaderfile:match("%.vert$") then
+            shader_type = "vertex"
+        elseif shaderfile:match("^fs_.*%.sc$") or shaderfile:match("%.frag$") then
+            shader_type = "fragment"
+        elseif shaderfile:match("^cs_.*%.sc$") or shaderfile:match("%.comp$") then
+            shader_type = "compute"
+        else
+            raise("cannot determine shader type from file name " .. path.filename(shaderfile))
+        end
+
         -- build command args
         local args = {
             "-f", shaderfile,
             "--type", shader_type,
             "--varyingdef", vardef_filename,
             "--platform", bgfx_platforms[target:plat()],
         }
+
+        if fileconfig and fileconfig.header then
+            table.insert(args, "--bin2c")
+            if fileconfig.header ~= true then
+                table.insert(args, fileconfig.header)
```

Comment:
I think there should be more info about `.header` argument in the comments at the beginning of this file, e.g. What would happen if both `-o <output>` and `--bin2c <header>` are specified? Which of the `<output>` and the `<header>` will be generated? Is `output_name = "shader.vert.h"` really correct?

Reply:
The `--bin2c` option generates a header file containing the compiled shader to be included in the project instead of loaded in runtime. It the output_name is defined, the ".h" suffix is recommended, but it just generates a header file. The "<header>" string is optional and, if provided it's used as the variable containing the shader data, if not defined it uses the input file base name. Since not all rule configuration values follow strictly the shaderc inputs I didn't follow also and used a more explicit name. What is the best way to document this rule, if it's necessary to read the source anyway, is that worth documenting it better?

Reply:
Of course it's better documenting the rule on some website like xrepo.xmake.io, but it's not yet implemented, so as for now the only way to document it is to do so in the rule file. Seems that when there is `--bin2c <header>`, the variable name in the generated file will be `<header>`? If so, I'd suggest a different name for this config, e.g. `export_var` or `array_name`(align with https://bkaradzic.github.io/bgfx/tools.html#shader-compiler-shaderc ) or just `array` to be more clear.

Reply:
`array_name=true` can become an advanced feature (deduce by extension .h is usually enough), so it's ok to document in the following way:
```
-- add_rules("@bgfx/shaders")
-- generate binary file:
-- add_files("shader.vert", {type = "vertex", output_dir = "shaders", output_name = "shader.vert.bin", profiles = {glsl = "330"}})
-- generate header file:
-- add_files("vs_shader.sc", {type = "vertex", output_dir = "shaders", output_name = "shader.vert.h", profiles = {glsl = "330"}})
-- generate header file exporting variable "vertex_src":
-- add_files("vs_shader.sc", {type = "vertex", output_dir = "shaders", output_name = "shader.vert.h", array_name = "vertex_src", profiles = {glsl = "330"}})
-- force to generate header file with default variable name:
-- add_files("vs_shader.sc", {type = "vertex", output_dir = "shaders", output_name = "shader.vert.inc", array_name = true, profiles = {glsl = "330"}})
```